### PR TITLE
Add history of validator's withdrawal credentials

### PR DESCRIPTION
**Pull request by samlaf**
This PR implements the changes proposed in https://github.com/wealdtech/chaind/issues/78#issuecomment-1423810954

We did not do extensive testing of these changes, but they seem to be working fine when indexing zhejiang. Let us know if there's anything we got wrong or that we can improve.

### CODE REVIEW

File: services/chaindb/postgresql/blstoexecutionchanges.go [hunk 1]
```diff
@@ -48,18 +48,21 @@ func (s *Service) setBLSToExecutionChanges(ctx context.Context, block *chaindb.B
 INSERT INTO t_block_bls_to_execution_changes(f_block_root
                                             ,f_block_number
                                             ,f_index
+											,f_validator_index
```
Comment (mcdee) [inline services/chaindb/postgresql/blstoexecutionchanges.go:51]: It looks like there are tabs used here; whitespace in the commands should be spaces.  Please could you correct here and elsewhere this shows up?

File: services/chaindb/postgresql/upgrader.go [hunk 2]
```diff
@@ -1496,12 +1511,6 @@ CREATE INDEX IF NOT EXISTS i_block_withdrawals_3 ON t_block_withdrawals(f_valida
 		return errors.Wrap(err, "failed to create block withdrawals index 3")
 	}
 
-	if _, err := tx.Exec(ctx, `
-CREATE INDEX IF NOT EXISTS i_block_withdrawals_4 ON t_block_withdrawals(f_address);
-`); err != nil {
-		return errors.Wrap(err, "failed to create block withdrawals index 4")
-	}
-
```
Comment (mcdee) [inline services/chaindb/postgresql/upgrader.go:1504]: Why is this index being removed?

File: services/chaindb/postgresql/upgrader.go [hunk 3]
```diff
@@ -1527,3 +1536,24 @@ CREATE TABLE t_fork_schedule (
 
 	return nil
 }
+
+func recreateValidators(ctx context.Context, s *Service) error {
```
Comment (mcdee) [inline services/chaindb/postgresql/upgrader.go:1540]: This doesn't seem like a suitable name for the function, something like `addValidatorWithdrawalCredentials` would be clearer.

**Review by mcdee — COMMENTED** [review]
Thank you for this PR.  A few items to change, mainly around formatting.

File: services/chaindb/postgresql/upgrader.go (hunk 3 above)
Comment (samlaf) [reply services/chaindb/postgresql/upgrader.go]: Changed

File: services/chaindb/postgresql/upgrader.go (hunk 2 above)
Comment (samlaf) [reply services/chaindb/postgresql/upgrader.go]: Not sure.. probably a mistake. Readded it.

File: services/chaindb/postgresql/blstoexecutionchanges.go (hunk 1 above)
Comment (samlaf) [reply services/chaindb/postgresql/blstoexecutionchanges.go]: Fixed

**Comment by samlaf** [comment]
@mcdee should be good to go.

**Review by mcdee — CHANGES_REQUESTED** [review]
Please could you rebase this against master so we can run the testing and linting?  There is also one further item that needs to be addressed.

File: services/chaindb/types.go [hunk 4]
```diff
@@ -51,6 +51,7 @@ type Validator struct {
 	ActivationEpoch            phase0.Epoch
 	ExitEpoch                  phase0.Epoch
 	WithdrawableEpoch          phase0.Epoch
+	WithdrawalCredentials      []byte
```
[the remark this answers is not in the document]
Comment (samlaf) [reply services/chaindb/types.go]: Done.

**Comment by samlaf** [comment]
hmm I might have screwed the commit history (hopefully only that..) by rebasing and force-pushing to my repo.
Your requested changes are done though, this should be good to go.

**Review by mcdee — APPROVED** [review]

**Comment by mcdee** [comment]
There is one outstanding issue picked up by linting, please could you address that?  After that it should be good to go.

**Comment by mcdee** [comment]
I'll merge this and fix it up, to allow running against sepolia.

**Comment by samlaf** [comment]
My bad was busy this morning. Thank you for fixing this.

**Pull request merged**